Fix #14572: Improve explaination for using `RUN` command inside Docke…

charleswhchan opened the pull request:
…rfile and added example.

Based on example provided by @thaJeztah:
  https://github.com/docker/docker/issues/14572#issuecomment-131469695

Signed-off-by: Charles Chan charleswhchan@users.noreply.github.com

thaJeztah commented on docs/articles/dockerfile_best-practices.md:
```diff
@@ -146,16 +146,33 @@ with backslashes.
 maintainers. If you know there’s a particular package, `foo`, that needs to be
 updated, use `apt-get install -y foo` and it will update automatically.
 
+* Don’t use `RUN apt-get update` on a single line. This will cause
```
I see this was already like this, but I think we should avoid "on a single line", because readers might confuse that with "you should wrap this line".

Thinking what would be a better way to describe this, perhaps something with "Don't put `apt-get update` in a separate `RUN` statement".

I'd like to defer this to @moxiegirl though. :relaxed:

thaJeztah commented:
Thanks so much again, @charleswhchan! I think this information is useful and clarifies *why* this shouldn't be done.

@moxiegirl recently explained to me that it's better to avoid "future tense" in technical documentation, and, for example, "all layers will be in the cache" would better be written as "all layers are in the cache". I think it's best if she has a look at this (I can make suggestions, but I'll probably mess that up :smile_cat:)

In general "LGTM" as it's an improvement, but;

ping @moxiegirl if you have some time, perhaps you can wave your ✨magic✨ wand, or LGTM

charleswhchan commented:
@thaJeztah: Thanks for the excellent feedback and teachings.

I will aait for additional input from @moxiegirl. :smiley:

moxiegirl commented on docs/articles/dockerfile_best-practices.md:
```diff
@@ -146,16 +146,33 @@ with backslashes.
-caching issues if the referenced archive gets updated, which will make your
-subsequent `apt-get install` fail without comment.
-
 * Avoid `RUN apt-get upgrade` or `dist-upgrade`, since many of the “essential”
```
Thanks @charleswhchan  for the contribution. Most of the `RUN` material is taken up with a discussion of `apt-get` which is the real topic here.  It would be better to give a heading to `apt-get` and raise its importance. This goes beyond the scope of your original change but if you are up to expanding a bit your help and technical review is invaluable.

---> https://gist.github.com/moxiegirl/06f680798666910c9749

### RUN

[Dockerfile reference for the RUN instruction](https://docs.docker.com/reference/builder/#run)

As always, to make your `Dockerfile` more readable, understandable, and
maintainable, put long or complex `RUN` statements on multiple lines separated
with backslashes.


### apt-get 

Probably the most common use-case for `RUN` is an application of `apt-get`. The
`RUN apt-get` command, because it installs packages, has several gotchas to look
out for.

You should avoid `RUN apt-get upgrade` or `dist-upgrade`, as many of the
“essential” packages from the base images won't upgrade inside an unprivileged
container. If a base package is out-of-date, you should contact its maintainers.
If you know there’s a particular package, `foo`, that needs to be updated, use
`apt-get install -y foo` to update automatically.

Always combine  `RUN apt-get update` with `apt-get install` in the same `RUN`
statement, for example:

        RUN apt-get update && apt-get install -y \
            package-bar \
            package-baz \
            package-foo


Using `apt-get update` alone causes caching issues and subsequent `apt-get
install` instructions fail. For example, say you have a Dockerfile:

        FROM ubuntu:14.04
        RUN apt-get update
        RUN apt-get install -y curl

After building the image, all layers are in the Docker cache. Suppose you later
modify `apt-get install` by adding extra package:

        FROM ubuntu:14.04
        RUN apt-get update
        RUN apt-get install -y curl nginx

Docker sees the initial and modified instructions as identical and reuses the
cache from previous steps. As a result the `apt-get update` is *NOT* executed
because the build uses the cached version. Because the `apt-get update` is not
run, your build can potentially get an outdated version of the `curl` and `nginx`
packages.

Using  `RUN apt-get update && apt-get install -y` ensures your Dockerfile
installs the latest package versions without no further coding or manual
intervention. This technique is known as "cache busting." You can also achieve
cache-busting by specifying a package version. This is known as version pinning,
for example:

        RUN apt-get update && apt-get install -y \
            package-bar \
            package-baz \
            package-foo=1.3.*

Version pinning forces the build to retrieve a particular version regardless of
what’s in the cache. This technique can also reduce failures due to unanticipated changes
in required packages.

Below is a well-formed `RUN` instruction that demonstrates the all the `apt-get`
recommendations. 

    RUN apt-get update && apt-get install -y \
        aufs-tools \
        automake \
        build-essential \
        curl \
        dpkg-sig \
        libcap-dev \
        libsqlite3-dev \
        lxc=1.0* \
        mercurial \
        reprepro \
        ruby1.9.1 \
        ruby1.9.1-dev \
        s3cmd=1.1.0*

The `s3cmd` instructions specifies a version `1.1.0*`. If the image previously
used an older version, specifying the new one causes a cache bust of `apt-get
update` and ensure the installation of the new version. Writing on package on
each line can also prevent mistakes in package duplication.

thaJeztah replied:
Looks good! Spotted a tiny typo while reading (haven't given it a thorough read);

> installs the latest package versions without no further coding or manual

"without no" -> "with no"

Should we add a "post install cleanup" to the "full" example? This is what my standard blocks look like;

```Dockerfile
RUN apt-get update && apt-get install -y \
    foo \
    bar \
    baz \
 && apt-get clean \
 && rm -r /var/lib/apt/lists/*
```

moxiegirl commented:
@thaJeztah If we add that, maybe mention the benefits of why we would recommend the clean / rm.

moxiegirl commented:
Disabling trust is discussed in the overview here: 

http://docs.docker.com/security/trust/content_trust/#disable-content-trust-for-specific-operations

thaJeztah commented:
@moxiegirl cleaning up the apt cache keeps the image-size down. Besides (if you're following the best practices explained), each `apt-get install` has its own `apt-get update`, so basically the cache is redundant.

From the [apt-get howto](https://help.ubuntu.com/community/AptGet/Howto);

> ... removes all packages from the package cache. This may not be desirable if you have a slow Internet connection, since it will cause you to redownload any packages you need to install a program.

Clearing out `/var/lib/apt/lists` can be a big saver in disk-space (see, e.g. http://askubuntu.com/questions/179955/var-lib-apt-lists-huge-in-12-04), it's also not needed, because running `apt-get update` will refresh that data

charleswhchan commented:
@moxiegirl: Sorry, I am late to join the discussions for https://gist.github.com/moxiegirl/06f680798666910c9749 ... a few suggestions?

> put long or complex RUN statements on multiple lines separated
with backslashes.

[1] or "split long or complex ..."

> If a base package is out-of-date, you should contact its maintainers.

[2] "If a package contained in the base image is out-of-date, you should contact its maintainers"

> Using `apt-get update` alone causes caching issues ...

[3] "Using `apt-get update` alone in a `RUN` statement causes caching issues ...

> "cache busting."

[4] "cache busting". (period after double quotes)

(btw: nice tip to use version pinning :+1: )

> Below is a well-formed `RUN` instruction that demonstrates the all the `apt-get` recommendations.

[5] ... that demonstrates --the-- all the ....

> s3cmd=1.1.0*

[6] Not sure if this is a typo? Should it be 1.1* or 1.1.* instead? 

> Writing on package on each line can also prevent mistakes in package duplication.

[7] Listing --on-- packages on each line and in alphabetical order also prevents mistakes in package duplication.

charleswhchan commented:
@thaJeztah: Is `apt-get autoremove` necessary for "post install cleanup"? 

I suppose there will not free up too much space than `clean` and is therefore less significant.

moxiegirl commented:
@charleswhchan I'm happy with your suggestions. Do you want to make them in combination with the gist and rebase?

charleswhchan commented:
@moxiegirl: Sure, will integrate the changes and update PR.

charleswhchan committed:
Fix #14572: Improve explaination for usng `RUN` and `apt-get` inside Dockerfil.

Based on example provided by @thaJeztah and @moxiegirl:
  https://github.com/docker/docker/pull/15773

Signed-off-by: Charles Chan <charleswhchan@users.noreply.github.com>

charleswhchan commented:
Rebase from upstream/master, and integrate gist and suggestions. 

Also added @thaJeztah's `apt-get clean` and `/var/lib/apt/lists` suggestion with a brief explaination.

SvenDowideit commented:
nice - late night LGTM from me :)

moxiegirl commented:
Consensus @charleswhchan you rock!  LGTM for the merge.

charleswhchan commented:
Thanks to everyone for all the help I got. :smile:

thaJeztah commented:
Thanks @charleswhchan for another great contribution!
